### PR TITLE
Don't convert id property to ObjectId if type is a String

### PR DESCRIPTION
**Pull request by wpjunior**
Don't convert id property to ObjectId if type is a String

### CODE REVIEW

**Comment by bajtos** [comment]
@raymondfeng ping

**Comment by bajtos** [comment]
@raymondfeng ping, how to move this patch forward?

**Comment by alFReD-NSH** [comment]
This change will fix #128 and strongloop/loopback#1441.

**Comment by slnode** [comment]
Can one of the admins verify this patch? To accept patch and trigger a build add comment ".*ok\W+to\W+test.*"

**Comment by ahmed-abdulmoniem** [comment]
Any news on this also?

**Comment by slnode** [comment]
Can one of the admins verify this patch? To accept patch and trigger a build add comment ".*ok\W+to\W+test.*"

**Comment by bajtos** [comment]
@0candy @jannyHou @Amir-61 could you PTAL?

**Comment by slnode** [comment]
Can one of the admins verify this patch?

**Comment by slnode** [comment]
Can one of the admins verify this patch?

**Comment by slnode** [comment]
Can one of the admins verify this patch?

File: lib/mongodb.js [hunk 1]
```diff
@@ -85,7 +85,6 @@ function MongoDB(settings, dataSource) {
   if (this.debug) {
     debug('Settings: %j', settings);
   }
-
```
Comment (Amir-61) [inline lib/mongodb.js]: Please don't touch unrelated code

File: lib/mongodb.js [hunk 2]
```diff
@@ -188,11 +187,12 @@ MongoDB.prototype.create = function (model, data, callback) {
   }
   var idValue = self.getIdValue(model, data);
   var idName = self.idName(model);
+  var idType = self.idType(model);
 
   if (idValue === null) {
     delete data[idName]; // Allow MongoDB to generate the id
   } else {
-    var oid = ObjectID(idValue); // Is it an Object ID?
+    var oid = idType(idValue); // Is it an Object ID?
```
Comment (Amir-61) [inline lib/mongodb.js]: ditto; please see https://github.com/strongloop/loopback-connector-mongodb/pull/67/files#r86277509

File: lib/mongodb.js [hunk 3]
```diff
@@ -233,8 +233,9 @@ MongoDB.prototype.save = function (model, data, callback) {
   }
   var idValue = self.getIdValue(model, data);
   var idName = self.idName(model);
+  var idType = self.idType(model);
 
-  var oid = ObjectID(idValue);
+  var oid = idType(idValue);
```
Comment (Amir-61) [inline lib/mongodb.js]: ditto

File: lib/mongodb.js [hunk 4]
```diff
@@ -262,7 +263,9 @@ MongoDB.prototype.exists = function (model, id, callback) {
   if (self.debug) {
     debug('exists', model, id);
   }
-  id = ObjectID(id);
+  var idType = self.idType(model);
```
Comment (Amir-61) [inline lib/mongodb.js]: ditto; `self.idType` is used only once in this block, you do not need to have this line... instead use it like:
```
id = self.idType(id);
```

File: lib/mongodb.js (hunk 2 above)
Comment (Amir-61) [inline lib/mongodb.js]: ditto

File: lib/mongodb.js (hunk 3 above)
Comment (Amir-61) [inline lib/mongodb.js]: ditto

File: lib/mongodb.js [hunk 5]
```diff
@@ -284,6 +287,7 @@ MongoDB.prototype.find = function find(model, id, callback) {
   }
   var idValue = id;
   var idName = self.idName(model);
+  var idType = self.idType(model);
```
Comment (Amir-61) [inline lib/mongodb.js]: ditto

File: lib/mongodb.js [hunk 6]
```diff
@@ -296,7 +300,7 @@ MongoDB.prototype.find = function find(model, id, callback) {
       }
     }
   }
-  var oid = ObjectID(idValue);
+  var oid = idType(idValue);
```
Comment (Amir-61) [inline lib/mongodb.js]: ditto

File: lib/mongodb.js [hunk 7]
```diff
@@ -323,8 +327,9 @@ MongoDB.prototype.updateOrCreate = function updateOrCreate(model, data, callback
 
   var id = self.getIdValue(model, data);
   var idName = self.idName(model);
+  var idType = self.idType(model);
```
Comment (Amir-61) [inline lib/mongodb.js]: ditto

File: lib/mongodb.js [hunk 8]
```diff
@@ -353,7 +358,9 @@ MongoDB.prototype.destroy = function destroy(model, id, callback) {
   if (self.debug) {
     debug('delete', model, id);
   }
-  id = ObjectID(id);
+  var idType = self.idType(model);
```
Comment (Amir-61) [inline lib/mongodb.js]: ditto

File: lib/mongodb.js [hunk 9]
```diff
@@ -463,12 +471,13 @@ MongoDB.prototype.all = function all(model, filter, callback) {
   }
   filter = filter || {};
   var idName = self.idName(model);
+  var idType = self.idType(model);
```
Comment (Amir-61) [inline lib/mongodb.js]: Is idType used anywhere in this block? It seems an unused variable

File: lib/mongodb.js [hunk 10]
```diff
@@ -604,8 +613,9 @@ MongoDB.prototype.updateAttributes = function updateAttrs(model, id, data, cb) {
   if (self.debug) {
     debug('updateAttributes', model, id, data);
   }
-  var oid = ObjectID(id);
   var idName = this.idName(model);
+  var idType = this.idType(model);
+  var oid = idType(id);
```
Comment (Amir-61) [inline lib/mongodb.js]: Remove `var idType = this.idType(model);`
and just:
```
 var oid = self.idType(id);
```

File: test/mongodb.test.js [hunk 11]
```diff
@@ -126,11 +171,11 @@ describe('mongodb', function () {
       should.not.exist(err);
       should.not.exist(person.id);
       person._id.should.be.equal(3);
-      
+
```
Comment (Amir-61) [inline test/mongodb.test.js]: If you rebase your code on top of master all of these linting issues are already fixed.

File: test/mongodb.test.js [hunk 12]
```diff
@@ -76,6 +76,51 @@ describe('mongodb', function () {
...
+      });
+    });
+
+    after(function(done) {
+      CustomUser.destroyAll(function() {
+        done();
+      });
+    });
+
+    it('should to store id in string format', function(done){
```
Comment (Amir-61) [inline test/mongodb.test.js]: `stores id in string format`

File: test/mongodb.test.js [hunk 13]
```diff
@@ -76,6 +76,51 @@ describe('mongodb', function () {
...
+        if (err) { return done(err); }
+        should.not.exist(err);
+        should.exist(data[0]);
+
+        data[0]._id.should.not.be.an.instanceOf(db.ObjectID);
+        done();
+      });
+    });
+
+    it('should find by string', function(done){
```
Comment (Amir-61) [inline test/mongodb.test.js]: `finds by string`

File: test/mongodb.test.js [hunk 14]
```diff
@@ -76,6 +76,51 @@ describe('mongodb', function () {
...
+      CustomUser.findById('54637b427ff88a00246e1154', function(err, item) {
+        if (err) { return done(err); }
+        should.not.exist(err);
```
Comment (Amir-61) [inline test/mongodb.test.js]: you have already returned by call back if `err` exists. no needs of this line

File: test/mongodb.test.js [hunk 15]
```diff
@@ -76,6 +76,51 @@ describe('mongodb', function () {
...
+        should.exist(data[0]);
+
+        data[0]._id.should.not.be.an.instanceOf(db.ObjectID);
+        done();
+      });
+    });
+
+    it('should find by string', function(done){
+      CustomUser.findById('54637b427ff88a00246e1154', function(err, item) {
+        if (err) { return done(err); }
```
Comment (Amir-61) [inline test/mongodb.test.js]: We follow this convention in LoopBack:
```
if (err) return done(err);
```

**Comment by Amir-61** [comment]
@bajtos 
>@0candy @jannyHou @Amir-61 could you PTAL?

I added a bunch of comments on the code; I leave the rest to @0candy @jannyHou; you can take it from here.

@wpjunior Your PR seems pretty old; it needs to be rebased on top of master and lots of linting issues are already fixed.

File: test/mongodb.test.js [hunk 16]
```diff
@@ -76,6 +76,51 @@ describe('mongodb', function () {
...
+    });
+
+    after(function(done) {
+      CustomUser.destroyAll(function() {
+        done();
+      });
+    });
+
+    it('should to store id in string format', function(done){
+      var cursor = db.connector.db.collection('CustomUser').find({_id: '54637b427ff88a00246e1154'});
```
Comment (jannyHou) [inline test/mongodb.test.js]: Do we have to test `cursor` here? 
I think we should test
```js
CustomUser.find({id: '54637b427ff88a00246e1154'}, function(err, result) {
  // typeof result.id should be `String`
})
```

File: lib/mongodb.js [hunk 17]
```diff
@@ -821,4 +831,6 @@ MongoDB.prototype.ping = function (cb) {
   }
 };
 
-
+MongoDB.prototype.idType = function(model) {
```
Comment (jannyHou) [inline lib/mongodb.js]: Sorry I am confused with the function `idType`, does it "return the idType of a model" or "return the converted id value in a certain type"?

From the function definition it takes in the model as param then satisfies "return the idType of a model"
But from the usage like [`var oid = idType(idValue);`](https://github.com/wpjunior/loopback-connector-mongodb/blob/b474da06a06606b914d9d6ad87a860c9f33df714/lib/mongodb.js#L190) you want it to "return the converted id value in a certain type".

Could you elaborate more? Thanks :)

**Comment by jannyHou** [comment]
A quick confirm of the bug fixing here:
```js
MyModel.find({id: 'str_value'}, function(err, result) { 
  // typeof result.id should be string.
  // but now it is ObjectId
});
```
cc: @raymondfeng @bajtos @Amir-61 

@wpjunior Could you take a look of my comments and apply feedback? Thank you:

- Rewrite test case: https://github.com/strongloop/loopback-connector-mongodb/pull/67#discussion_r86834634
- Misuse of function `idType()`: https://github.com/strongloop/loopback-connector-mongodb/pull/67#discussion_r86841546

And in my opinion the id conversion should take place in function [`MongoDB.prototype.toDatabase`](https://github.com/strongloop/loopback-connector-mongodb/blob/master/lib/mongodb.js#L229), not necessarily touching functions like `create`, ect... Correct me if I am wrong :) Thanks.

**Comment by b-admike** [comment]
@wpjunior thank you for your contribution. Can you please apply the review comments you received above and rebase your branch against master? I believe this PR could fix #363

**Comment by slnode** [comment]
Can one of the admins verify this patch?

**Comment by b-admike** [comment]
@wpjunior I'm going to close this PR because it has been stale for a long time, and is not particularly related to any issues. We can revisit it later if necessary. Thank you again.
